cmd/snap-confine/mount-support.c

```diff
@@ -671,8 +672,20 @@ void sc_populate_mount_ns(struct sc_apparmor *apparmor, int snap_update_ns_fd,
 	}
 
 	// set up private mounts
+	if (getegid() != 0 && saved_gid == 0) {
+		// Temporarily raise egid so we can create, chmod and chown
+		// the mount without causing a noisy fsetid capability denial
+		if (setegid(0) != 0) {
+			die("cannot set effective group id to root");
+		}
+	}
 	// TODO: rename this and fold it into bootstrap
 	setup_private_mount(inv->snap_instance);
+	if (geteuid() == 0 && real_gid != 0) {
```

Comment:
should this be `geteuid()` or `getegid()` ?

Reply:
In terms of outcome, it doesn't matter since:
 * if 6755 and non-root, enter this function as ruid=1000,euid=0,suid=0,rgid=1000,egid=1000,sgid=0 so we raise (since egid=1000 and sgid=0) and drop (since euid=0 (but so is egid since we just raised) and rgid=1000)
 * if root, enter this function as ruid=0,euid=0,suid=0,rgid=0,egid=0,sgid=0 and as desired, don't drop since egid=0 and don't raise since ruid=0

This code follows the pattern established elsewhere for raise/drop which provides for easier review.

Honestly, snap-confine has grown rather organically and as such could use a little refactoring in this area so it is clearer (something along the lines of what I suggested in https://github.com/snapcore/snapd/pull/7625#discussion_r339823092 where we permanently drop immediately/use 4755 permissions (ie, ruid=1000,euid=0,suid=0,rgid=1000,egid=1000,sgid=1000) then have a utility function that safely creates dirs (eg, sc_mkdir(path, uid, gid, 0700). In this manner, it is relatively easy to reason about the permissions at any given moment and any group raise/drop happens in precisely one spot (sc_mkdir()).

Reply:
thanks for the in-depth explanation!

Reply:
FYI, I did some pseudo-code for what I was thinking for cleaning this up here: https://github.com/snapcore/snapd/pull/7625#discussion_r355010063